### DISCUSSION TITLE
json patch syntax for overlays

### DISCUSSION THREAD

**Discussion by LukeWinikates** [k14s/ytt#313] [Ideas]
I work on some projects that use `ytt`, and some that use `kustomize`.

For patching array items, I do like the use of `json path` and `json patch` that `kustomize` supports.

https://github.com/kubernetes-sigs/kustomize/blob/master/examples/jsonpatch.md

`ytt`'s matcher functions give a nice option for when you have more complex needs, but I wonder if an `@overlay/match by=overlay.jsonpath(/foo/bar/2/baz)` option would be consise and readable in some cases. It seems to me that `ytt` overlays require several nested layers of `#@overlay/match` for arrays within arrays.

**Comment by pivotaljohn** [comment]
I like the idea of using a notation that is likely familiar to the person using the tool. [`kubectl` also accepts JSON path expressions](https://kubernetes.io/docs/reference/kubectl/jsonpath/) to shape query results — though, notably, the user is redirected to a family of flags and/or `jq` for actually filtering results, themselves.

**Why Structures?**
An aspect of `ytt`'s design is that it encourages the practitioner to think and operate in terms of structures.

Granted that this almost always a shift in thinking. We're painfully aware of the pain that many folks go through to make that shift. However, we believe that given the time to do so, we can mitigate those pains sufficiently to make the host of benefits on the other side well worth that trade-off. I don't want to distract with elaborating the total scope of those benefits, here, but just consider what can be accomplished when the tool has information about the exact shape (and potentially meaning) of the structure.

One relevant benefit to the user can be increasing their confidence while visually reasoning that the path they have described will match the documents they intend. For example, one trick is when writing an overlay, I can copy the relevant snippet from an actual document to seed my overlay and then resolve the ambiguities with annotations. Modulo the sharp edges around overlays (of which many are getting filed down, as we type), this is a great way to "get it right" out the gate.

An indirect relevant benefit to the user is that there's one mechanism/symbol-system for talking about documents: YAML. There's one notation to learn (albeit, an expanding notation when doing things like patching): almost all of what you know about the notation works as you take on more capabilities. This also keeps the [implementation and interface complexity](http://www.catb.org/~esr/writings/taoup/html/ch13s01.html) of the tool itself minimized. Tools that do this tend to mature more quickly and have fewer defects: they are more reliable.

...

So, if there seems to be resistance to supporting JSON Path notation, well ... it can be in part because doing so — despite the obvious value of reuse across (some) tools in the ecosystem — it would mitigate these desired qualities of the tool.

Now, it's not to say that once the Carvel tools reach a certain level of maturity that they couldn't be in a place to take on convenience features like supporting JSON Path. Looking into _my_ 🔮 , I see that once this tool starts to tackle problems that practically speaking require paths, it would be important to seriously consider this notation.

All of this said to expose the structure of some thinking around the topic. Others (even from within the project) have different perspectives. And impact does not necessarily follow intention.
We'd be delighted to hear about gaps, failures, and reports from our blind spots; the ultimate goal is to Do the Right Thing.

**Reply by LukeWinikates** [reply]
Ok, so - I admit that I am skeptical of the "it's yaml, so you already know it" idea. For me, the main learning hurdle with `ytt` overlays has been the `#@overlay/match` operations that I have to learn and include at what feel for me like arbitrary layers in the yaml that I'm overlaying. I'm not saying that I got my first `kustomize` JSON patch right on the first try, but In the same way that one of the most reassuring things about `ytt` is that it uses `starlark`, one of the reassuring things about `kustomize` is that JSON patch is a published standard.

I get that if I have two yamls, I can diff them against one another visually, but I don't think that JSON path/patch notation is any less "structural" - the hierarchical path notation is concise, and refers to individual nodes based on their location in the structure of the document. The `#@overlay/match` syntax and the need to specify it at many layers in a document is structural, but verbose - I also like that JSON patch very explicitly surfaces the nature of the change that it's trying to do via the "op" field. I find this harder to infer when I look at `ytt` overlays.

```
#@overlay/match by=overlay.subset({"kind":"Package"})
---
spec:
  template:
    spec:
      fetch:
        #@overlay/match by=overlay.all
        - imgpkgBundle:
            #@overlay/match missing_ok=True
            image: #@ data.values.dev_image
```

vs. `kustomize`

```
patchesJSON6902:
- target:
    kind: Package
  patch: |-
    - op: replace
      path: /spec/template/spec/fetch/0/imgpkgBundle/image
      value: #@ data.values.dev_image
```

or

```
#@overlay/match by=overlay.subset({"kind":"Package"})
---
#@overlay/match by=overlay.path("/spec/template/spec/fetch/0/imgpkgBundle/image")
replace: #@ data.values.dev_image
```

Maybe there's an unresolvable issue that JSON path syntax just doesn't fit nicely into the context of `ytt`, although the above represents an attempt on my part to sketch a syntax. I appreciate that y'all have a philosophy for steering `ytt's` features, but it's a little patronizing to read things like "this is almost always a shift in thinking", or to be linked to a document about managing software complexity, as though I am a student needing instruction rather than a peer trying to get my own work done. I am sure I'm also at fault for throwing a very abstract suggestion over the wall, somewhat expecting that it would be philosophically unwelcome. But I do find the`ytt` matcher syntax less convenient than json patch.

**Reply by pivotaljohn** [reply]
_(Apologies for the long delay in responding; I missed your initial reply and just saw this on Friday afternoon.)_

I am really sorry for making you feel disrespected, Luke. It's not how we do it, here. We want everyone in this community to be able to share their perspective and experience on an equal footing.

I absolutely can see how "Granted that this almost always a shift in thinking" can mean, "when folks finally come around to our way of thinking" effectively dismissing all differing views, out of hand. I also see that linking to Eric Raymond's book's definition of interface and implementation complexity can draw the inference that someone needs to be taught… and about some fundamentals — that's pretty condescending. That sucks and I'm sorry.

To lay it all bare, I meant no such disrespect. By "Granted that this almost always a shift in thinking" I _meant_ to acknowledge that `ytt` currently places what often feels like high demands on the user... intending an empathetic nod. And the link to The Art of Unix Programming section on complexity was my attempt at mitigating some gatekeeper-y effect of using meaning-laden terms… that by providing a reference, those who didn't know (presumably others reading our public conversation) could better follow the principle I was invoking.

So, I missed my own mark of "unambiguously inclusive" and I'm sorry you're the recipient of it. 

Thank you for your candor, Luke. Sharing sentiments in text — a medium devoid of almost all non-verbal elements — is risky. Given that my role in this exchange is to respectfully contextualize the idea you're bringing, I've got some work to do to more skillfully mitigate that risk.

**Reply by pivotaljohn** [reply]
Into the technical topic at hand: introducing RFC 6902-like syntax to `ytt` overlaying...

The heart of the message I'm reading, here is:
- `ytt` overlays require significant visual scanning vs. JSON patches are visually more compact and intention revealing.
- `ytt` overlays including arrays are verbose and it's unclear why some of those annotations are required.
- JSON patch is a standard, `ytt` overlays are not.
- overall the `ytt` overlay syntax feels less intuitive and therefore more taxing to read and write (i.e. get it right).
- given all this, why not just support JSON path notation _alongside_ other matchers?


### JSON patch: more compact paths
JSON Path is more concise than the corresponding multi-line YAML equivalent. Agreed: I too find it often is.

Though, turning this example over in my hand, I see that the `ytt` version is "only" 2 lines longer… if the target has fewer parent nodes, it could actually be shorter… and when there are multiple operations going on in one overlay, it can actually turn out to be fewer lines overall.

But I think what you're pointing at is more about the path _itself_. That in the JSON path you only need to scan _within_ one (1) line to get a feel for where the change is happening.


### JSON patch: more explicit intent

`op` makes explicit the intended edit. I see what you mean: one place, one verb.

Ironically, the implicitness in this `ytt` overlay example comes from it exclusively using the default operation: merge. Had we used another action (e.g. `@overlay/replace` or `@overlay/remove`), it _would_ be a bit more explicit.

Again, though, if I'm understanding what you're saying, it's not about whether the information is present, but that the eye doesn't have to scan over the whole YAML structure to find it.

Slight tangent: I notice that the two examples here are subtly different:
If the intent is to truly "replace" (for those reading along at home: [RFC 6902 § 4.3](https://tools.ietf.org/html/rfc6902#section-4.3)), then the `image` map item in the `ytt` overlay would be annotated: `@overlay/replace` (instead of `@overlay/match missing_ok=True`)
if the intent is to do a "replace or append", then the JSON patch would use the `add` operation (see [RFC 6902 § 4.1](https://tools.ietf.org/html/rfc6902#section-4.1)).


### Overlaying Arrays

Yeah, I have found overlay'ing arrays to be my own biggest sources of pain while using `ytt` (I started with the ~k14s~Carvel tools during a stint on the CF for K8s team). You can see evidence of my own early headbanging in vmware-tanzu/carvel-ytt#94.

One insight that helped things make more sense for me is: the array item on the "right" is evaluated many times: once for _each_ item in the array on the "left" (if needed, precise definitions of "left" and "right" are found in the docs: [Overlay Module > Overview](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-overlay/#overview)). 

Here, `ytt` overlays can do something that JSON patches cannot: select on array items based on their content, not just their index.

In fact, the example illustrates this well:

```yaml
#@overlay/match by=overlay.subset({"kind":"Package"})
---
spec:
  template:
    spec:
      fetch:
        #@overlay/match by=overlay.subset({"imgpkgBundle": {}})
        - imgpkgBundle:
            #@overlay/match missing_ok=True
            image: #@ data.values.dev_image
```

This shifts the connaissance from position to name (for those who haven't stumbled over the late (and great) Jim Weirick's treatment, I've found the concept useful in wrangling various forms of coupling: https://connascence.io).

Practically speaking, this means that there can be multiple `fetch` directives of which only the `imgpkgBundle` is affected by this edit.


### JSON paths in `ytt` Overlays?

All this said, what Lisp does for Emacs, Starlark does for `ytt`. Effectively [`@overlay/replace`](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-overlay/#overlayreplace) provides all that's needed to edit.

Poking at this strategy, with [this toy example](https://carvel.dev/ytt/#gist:https://gist.github.com/jtigger/633f308de6233e747ae2702100250e4a), I noticed:
- since the overlay reaches _into_ the left (rather than mirror it as the core mechanism is designed around) it requires merging the edit result back into the original YAML structure. Today, [YAML Fragments](https://carvel.dev/ytt/docs/latest/lang-ref-yaml-fragment) are immutable.
  - a hack is to launder the structure through the [`yaml` module](https://carvel.dev/ytt/docs/latest/lang-ref-ytt/#yaml)...
  - but this has the side-effect of washing away any meta (i.e. code, annotations, or type information) that were present.
  - perhaps neither here nor there since overlays apply on top of already-evaluated templates and type information is only being set on Data Value documents, right now. But it's a nasty landmine for sure: there's no warning.
  - would have to think about all the consequences of YAML Fragments _becoming_ mutable
- error messages are even more cryptic with Starlark details leaking out (e.g. get a [reference-token](https://tools.ietf.org/html/rfc6901#section-3) name wrong and the error messages start talking about `dict`s and `list`s).

Given all this, wondering what you're thinking, Luke.

I see the compact notation being appealing. There's the lack of selecting on array items other than index.  We want to keep our options open… and we have a history of including convenience features over time.

**Reply by LukeWinikates** [reply]
> The heart of the message I'm reading, here is:

Yep, that definitely captured my thoughts.

> But I think what you're pointing at is more about the path itself. That in the JSON path you only need to scan within one (1) line to get a feel for where the change is happening.

Yes, I think that's it - thank you for distilling that.

> Here, `ytt` overlays can do something that JSON patches cannot: select on array items based on their content, not just their index.

This might be true, but I'm not sure - `kustomize` is the only tool I've used that makes use of JSON patch, and I did array operations with it, but haven't used complex array filters. But they do seem to be in the JSON patch spec:

https://github.com/json-path/JsonPath#filter-operators

There seem to be limitations, so I would expect that there are things that you could do in `ytt` that you couldn't with `kustomize`, since `ytt` has a full embedded programming language and `kustomize` doesn't (by design, IIRC).

I am conflicted about overlays or patches that get this 'interesting' though. `ytt` seems to really lean into supporting complex overlays, but I admit that it makes me uneasy to use the more complex features - I think this is just how I feel about overlays, because I felt this unease using JSON patch to manipulate ingress rules with `kustomize` as well. It felt like it was right on the edge of what's reasonable to do using an overlay vs finding some other way (templating, bash scripting, refactoring the base yaml so that less complex overlay patterns could be viable, etc)

> Given all this, wondering what you're thinking, Luke.

Wow, thanks for building a proof of concept in gist form, and sharing the implementation challenges. I don't know - I don't have the maintainer's-eye view of the situation. I'm grateful for the concern about cryptic error messages bubbling up from the Starlark layer, and I don't have much of a suggestion there other than to strongly agree that error messages are really important because debugging overlays is hard. I think `ytt` does well right now to include stack traces for where the overlay matching goes wrong. The head start to figuring out where to try something else is helpful.

If I had to guess what made it frustrating for me to debug my own overlays, it was the breadth of the matching operation vocabulary and not knowing which of the different verbs to look for. Maybe that suggests that rather than having multiple idioms for identifying a particular yaml node and editing it (the existing `ytt` patterns + JSON patch), an investment it getting error messages to the point where `ytt` can tell you why your match rules didn't work and suggest a "maybe you meant:" alternative.

**Reply by pivotaljohn** [reply]
@LukeWinikates said:
> I am conflicted about overlays or patches that get this 'interesting' though. ytt seems to really lean into supporting complex overlays, but I admit that it makes me uneasy to use the more complex features - I think this is just how I feel about overlays, because I felt this unease using JSON patch to manipulate ingress rules with kustomize as well. It felt like it was right on the edge of what's reasonable to do using an overlay vs finding some other way (templating, bash scripting, refactoring the base yaml so that less complex overlay patterns could be viable, etc)

Oh yeah, this sentiment really resonates with me too. Overlays are multiple flavors of indirection happening at once: they are declarative ("did my overlay match on _exactly_ the set of documents I intended?"); and they are expressions ("aaand, what does the resulting YAML look like?!").

Whenever someone can capture customization in templating, many folks on the Carvel team will recommend doing so citing the inherent complexity of overlays.

Yeah, the expressive power is there as a means of last resort; for example, you don't own the YAML and you can't (practically) inject your own additional post-processing.


@LukeWinikates reflected:
> If I had to guess what made it frustrating for me to debug my own overlays, it was the breadth of the matching operation vocabulary and not knowing which of the different verbs to look for. Maybe that suggests that rather than having multiple idioms for identifying a particular yaml node and editing it (the existing ytt patterns + JSON patch), an investment it getting error messages to the point where ytt can tell you why your match rules didn't work and suggest a "maybe you meant:" alternative.

Yes!

We've had an initiative to specifically [identify ways to improve our Overlay error messaging](https://hackmd.io/team/carvel?nav=overview&tags=%5B%22improving-overlay-err-msgs%22%5D). This is an effort that — through prioritization — has had to yield to addressing straight-up missing features.

That said, your observation and suggestion here can fold into that synthesis:

What can be done to:
- adjust the verbs; and/or
- make clear "how" to articulate specific operations in an overlay... and
- bake that understanding back into the UX so that the combination of docs, examples, help text, error messages, (other?) will build up the user's mental model so that it's:
  - easier to "guess" how to say a thing
  - quicker to find my way back out of dead-ends
  - and therefore be more productive.
?


Earlier, @LukeWinikates pointed out:
> But they do seem to be in the JSON patch spec:
> 
> https://github.com/json-path/JsonPath#filter-operators
> 
> There seem to be limitations, so I would expect that there are things that you could do in ytt that you couldn't with kustomize, since ytt has a full embedded programming language and kustomize doesn't (by design, IIRC).

Yeah, that seems kinda perfect... just enough of language to articulate a predicate. 👍🏼 

FWIW, my research is showing that JSON Path is a different (evolved?!?) thing from JSON Pointer (i.e. RFC 6901). It's the latter that JSON Patch (RFC 6902) is based on — the same flavor of patching that [Kustomize is currently using](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/jsonpatch.md#json-patching).

[RFC 6902 § 4](https://tools.ietf.org/html/rfc6902#section-4) makes that explicit reference:
> _That member's value is a string containing a JSON-Pointer value [[RFC6901](https://tools.ietf.org/html/rfc6901)] that references a location within the target document (the "target location") where the operation is performed._

**Reply by LukeWinikates** [reply]
> Whenever someone can capture customization in templating, many folks on the Carvel team will recommend doing so citing the inherent complexity of overlays.

Yes... tangentially, this is coming up a lot with references to K8s namespaces. They seem like a place where overlays are natural, because they are an obvious place where the person apply the yaml will often want to decide where all the resources are deployed. But as we Carvelize yaml fragments from other sources, we often find references to the namespace in spots like DNS names in arbitrary spots, or most recently in a `podAffinityTerm`.

I think to some degree these references to the namespace can be avoided altogether so they don't need to be modified by an overlay. But this the example that I keep coming back to when it comes to questioning the sweet spot for overlays as well as what the ideal templating solution for K8s resource definitions should look like.

Appreciated your response about the overlay error messages. I'll have to keep thinking if I have any concrete suggestions here.

> Yeah, that seems kinda perfect... just enough of language to articulate a predicate. 👍🏼

Yup, exactly - and thanks for calling out that that's a different RFC and not what Kustomize actually ships 😅
